Lesson29/havva/app.js

```diff
@@ -0,0 +1,142 @@
+const loadMore = document.getElementById('loadMore');
+const errorMsg = document.getElementById('error');
+const favorites = document.getElementById('favorites-container');
+const favoritesBtn = document.getElementById('favoritesBtn')
+let nextPage;
+let allLoadedPhotos = []; // this will store all visible photos in the gallery 
+
+button.addEventListener('click', async () => {
+  gallery.innerHTML = '';
+  allLoadedPhotos = [];
+
+  const data = await fetchData(
+    URL + 'photos',
+    true,
+    'orientation=portrait&count=10',
+  );
+  gallery.classList.remove('hidden');
+
+  allLoadedPhotos = allLoadedPhotos.concat(data.photos)
+  renderGallery(allLoadedPhotos);
+  setNextPage(data.link);
+});
+
+loadMore.addEventListener('click', async () => {
+  const data = await fetchData(nextPage, false);
+  if (!data || !data.photos) return // checks if the same photo is loaded before
+
+  allLoadedPhotos = allLoadedPhotos.concat(data.photos);
+
+  renderGallery(allLoadedPhotos);
+  setNextPage(data.link);
+});
+
+favoritesBtn.addEventListener('click', () => {
+  // we do not need async here because standard page navigation doesn't require async or await
+  window.open('favorites.html', target = '_blank');
+});
+
+function renderGallery(photosToRender) { // clear the gallery
+  gallery.innerHTML = '';
+  createImages(photosToRender);
+}
+
+
+function createImages(photos) {
+  let favoritePhotos = JSON.parse(localStorage.getItem('favoritePhotos')) || [];
+
+  for (let i = 0; i < photos.length; i++) {
+    const img = document.createElement('img');
+    const likeButton = document.createElement('button');
+    const photoContainer = document.createElement('div');
+
+    likeButton.innerHTML = '&#9825';
+
+    if (favoritePhotos.some(photo => photo.id === photos[i].id)) {
+      likeButton.innerHTML = '&#9829';
+    } else {
+      likeButton.innerHTML = '&#9825';
+    }
+
+    likeButton.addEventListener('click', () => {
+      let freshFavorites = JSON.parse(localStorage.getItem('favoritePhotos')) || [];
+
+      let isAlreadyFavorite = freshFavorites.some(photo => photo.id === photos[i].id);
+      if (isAlreadyFavorite) {
+        // If Yes: REMOVE it (Keep everything that does NOT match this ID)
+        freshFavorites = freshFavorites.filter(photo => photo.id !== photos[i].id);
+        likeButton.innerHTML = '&#9825';
+      } else {
+        // If No: ADD it
+        freshFavorites.push(photos[i]);
+        likeButton.innerHTML = '&#9829';
+      }
+      localStorage.setItem('favoritePhotos', JSON.stringify(freshFavorites));
+    });
+    img.src = photos[i].urls.small;
+    img.alt = photos[i].alt_description;
+
+    photoContainer.appendChild(img);
+    photoContainer.appendChild(likeButton);
+    gallery.appendChild(photoContainer);
+  }
+}
+
+function setNextPage(link) {
+  const maybeNextPage = getNextRelUrl(link);
+  if (maybeNextPage) {
+    loadMore.classList.remove('hidden');
+    nextPage = getNextRelUrl(link);
```

Comment:
The function getNextRelUrl is called twice on line 89 and line 92 with the same argument. Just reuse the result from line 89 on line 92 instead of calling the function again: 


```suggestion
    nextPage = maybeNextPage;
```